### PR TITLE
[SHELL32][INCLUDE] Fix PathResolve for double-backslash

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-18080](https://jira.reactos.org/browse/CORE-18080), [CORE-15204](https://jira.reactos.org/browse/CORE-15204)

## Proposed changes

- Fix `shell32!PathResolve` and `shell32!PathQualifyExW` functions for double-backslash.

## Comparison

BEFORE:
before
There were some failures.

AFTER:
after
Successful.

AFTER:
after2
The icon is fixed.

### CODE REVIEW

File: dll/win32/shell32/wine/shellpath.c [hunk 1]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
-{
-    FIXME("GetShortPathNameAbsentW(%ls, %p, %ld): stub\n", pszLong, pszShort, cchShort);
-    StringCchCopyW(pszShort, cchShort, pszLong);
-    return lstrlenW(pszShort);
-}
-
-BOOL WINAPI IsLFNDriveW(LPCWSTR lpszPath);
-
-/* @unconfirmed */
+/* @implemented */
```
Comment (binarymaster) [inline dll/win32/shell32/wine/shellpath.c:149]: Maybe add SAL2 annotation here instead of just `@implemented`. Also it looks like this is Wine code (i.e. `#ifdef __REACTOS__` is needed).

Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:149]:
```txt
dll/win32/shdocvw             # Synced to WineStaging-4.18
dll/win32/shell32             # Forked at Wine-20071011
dll/win32/shfolder            # Synced to WineStaging-3.3
dll/win32/shlwapi             # Synced to WineStaging-3.3
```

`SHELL32` is already forked.

Comment (binarymaster) [reply dll/win32/shell32/wine/shellpath.c:149]: Ah ok, but interesting that there are some existing `__REACTOS__` guards in this file :slightly_smiling_face:

**Comment by HBelusca** [comment]
@katahiromz You may take the opportunity to also fix the test output string:
```
Path expected L"C:\\ReactOS\\bin\\testdir\\2PRONG.txt", was L"C:\\ReactOS\\bin\\testdir\\\\2PRONG.txtL"C:\\ReactOS\\bin\\testdir\\2PRONG.txt".
```
where two strings are repeated after the "was" .

File: dll/win32/shell32/wine/shellpath.c (hunk 1 above)
Comment (HBelusca) [reply dll/win32/shell32/wine/shellpath.c:149]: Dunno what's the deal with this `@implemented` tbh.

**Comment by katahiromz** [comment]
> two strings are repeated after the "was" .

I think these strings are same behaviour as Windows' one. They are corner cases.

File: dll/win32/shell32/wine/shellpath.c [hunk 2]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+    FixSlashesAndColonW(szTemp); /* every '/' --> '\' */
 
-    if (pszDir)
+    /* Build the root-like path on pszPath, and set pchTemp */
+    if (PathIsUNCW(szTemp)) /* UNC path: Begins with double backslash */
     {
-        cch = GetCurrentDirectoryW(_countof(szCurDir), szCurDir);
-        if (cch <= 0 || cch >= _countof(szCurDir) || !SetCurrentDirectoryW(pszDir))
-            pszDir = NULL;
+        pszPath[2] = UNICODE_NULL; /* Cut off */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:167]: I'm not sure this is always done. There should be an extra condition there...

File: dll/win32/shell32/wine/shellpath.c [hunk 3]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+    pchPathEnd = pszPath + MAX_PATH;
+
+    /* Start appending the path components of szTemp to pszPath. */
+    while (*pchTemp && pchPath < pchPathEnd)
     {
-        if (!GetFullPathNameW(szCopy, _countof(szRoot), szRoot, NULL))
-            goto Quit;
-        if (!GetShortPathNameW(szRoot, szCopy, _countof(szCopy)) &&
-            !GetShortPathNameAbsentW(szRoot, szCopy, _countof(szCopy)))
+        if (pchTemp[0] == L'.') /* Dot? */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:208]:
```suggestion
        /* Collapse any .\ and ..\ parts in the path */
        if (pchTemp[0] == L'.')
```

File: dll/win32/shell32/wine/shellpath.c [hunk 4]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+                bDots = TRUE;
+            }
+            else
+            {
+                /* Component '.' */
+                if (pchTemp[1] == UNICODE_NULL || pchTemp[1] == L'\\')
+                {
+                    bDots = TRUE;
+                }
+            }
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:226]:
```suggestion
            else if (pchTemp[1] == UNICODE_NULL || pchTemp[1] == L'\\')
            {
                /* Component '.' */
                bDots = TRUE;
            }
```

File: dll/win32/shell32/wine/shellpath.c [hunk 5]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+/* @implemented */
 static VOID WINAPI PathQualifyExW(LPWSTR pszPath, LPCWSTR pszDir, DWORD dwFlags)
 {
-    WCHAR szRoot[MAX_PATH], szCopy[MAX_PATH], szCurDir[MAX_PATH];
-    LPWSTR pch;
-    LONG cch;
-    BOOL bCheckLFN;
+    INT iDrive;
+    WCHAR szTemp[MAX_PATH], szRoot[MAX_PATH];
+    LPWSTR pchTemp, pchPath, pchPathEnd;
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:154]:
```suggestion
    PWCHAR pchTemp, pchPath, pchPathEnd;
```

File: dll/win32/shell32/wine/shellpath.c [hunk 6]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+                /* Component '.' */
+                if (pchTemp[1] == UNICODE_NULL || pchTemp[1] == L'\\')
+                {
+                    bDots = TRUE;
+                }
+            }
+
+            if (bDots) /* '..' or '.' ? */
+            {
+                /* Go to next component */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:230]:
```suggestion
            /* If a '..' or '.' was encountered, skip to next component */
            if (bDots)
            {
```

File: dll/win32/shell32/wine/shellpath.c [hunk 7]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+                while (*pchTemp && *pchTemp != L'\\')
+                {
+                    ++pchTemp;
+                }
+
+                if (*pchTemp == L'\\')
+                {
+                    ++pchTemp;
+                }
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:239]:
```suggestion
                    ++pchTemp;
```

File: dll/win32/shell32/wine/shellpath.c [hunk 8]
```diff
@@ -146,90 +146,136 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
-    LONG cch;
-    BOOL bCheckLFN;
+    INT iDrive;
+    WCHAR szTemp[MAX_PATH], szRoot[MAX_PATH];
+    LPWSTR pchTemp, pchPath, pchPathEnd;
 
-    if (FAILED(StringCchCopyW(szCopy, _countof(szCopy), pszPath)))
+    /* FIXME: Short pathname */
+
+    /* Copy path to szTemp */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:158]: A good comment explains why such a copy is done (instead of rerpeating what the name of the function called already tells).

**Comment by HBelusca** [comment]
> > two strings are repeated after the "was" .
> 
> I think these strings are the results of the same behaviour as Windows' one. They are corner cases.

In the test case, line 660, I'm not sure the 1st parameter can be NULL. It's not optional according to the docs, and is directly used without NULL check in the implementation of PathResolve:
```
 660             Ret = (*s_pPathResolve)(NULL, pEntry->Dirs, pEntry->Flags);
```

File: dll/win32/shell32/wine/shellpath.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: Deleted.

File: dll/win32/shell32/wine/shellpath.c [hunk 9]
```diff
@@ -675,79 +721,58 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
 
+    if (flags & PRF_DONTFINDLNK)
+        dwWhich &= ~WHICH_LNK; /* Don't search '.LNK' (shortcut) */
+
     if (flags & PRF_VERIFYEXISTS)
-        SetLastError(ERROR_FILE_NOT_FOUND);
+        SetLastError(ERROR_FILE_NOT_FOUND); /* We set this error code at first in verification */
 
-    PathUnquoteSpacesW(path);
+    PathUnquoteSpacesW(path); /* Unquote the path */
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: Deleted.

File: dll/win32/shell32/wine/shellpath.c (hunk 8 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: Improved.

File: dll/win32/shell32/wine/shellpath.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:169]: I don't support non-LFN at this moment.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_02_11_2022_05_09_13
OK.

**Comment by katahiromz** [comment]
> directly used without NULL check in the implementation of PathResolve:

That will be caught by SEH.

**Comment by HBelusca** [comment]
> > directly used without NULL check in the implementation of PathResolve:
> 
> That will be caught by SEH.

My question here basically was whether it was intentional to pass NULL in the first parameter, instead of in the second.

File: dll/win32/shell32/wine/shellpath.c (hunk 2 above)
Comment (HBelusca) [reply dll/win32/shell32/wine/shellpath.c:169]: Then adding a // FIXME / TODO comment could help...

Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:169]: There is that at Line 157.

Comment (HBelusca) [reply dll/win32/shell32/wine/shellpath.c:169]: Yes, but it doesn't tell where exactly to fill the code holes.

Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:164]: Fixed.

**Comment by katahiromz** [comment]
Yep, intentional.

**Comment by HBelusca** [comment]
OK then it was not at all clear in the code of the test (missing explanatory comment).

**Comment by HBelusca** [comment]
I'll finish to review the PR tomorrow.

**Comment by katahiromz** [comment]
This PR partially fixes [CORE-15204](https://jira.reactos.org/browse/CORE-15204):

BEFORE:
before-PR#4833

AFTER:
after-PR#4833

File: dll/win32/shell32/wine/shellpath.c [hunk 10]
```diff
@@ -673,38 +712,40 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && /* Try to find a program? */
+            PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich)) /* Search it */
+        {
+            return TRUE; /* Found */
+        }
 
-        if (PathFindOnPathW(path, dirs))
+        if (PathFindOnPathW(path, dirs)) /* Try to find the filename in the directories */
         {
 #if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:752]: This exists also on Win2k3 SP2 at least.

File: dll/win32/shell32/wine/shellpath.c [hunk 11]
```diff
@@ -713,41 +754,37 @@ BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
...
-                return FALSE;
-            }
+            return TRUE; /* Found */
         }
 
+        if (!PathFileExistsAndAttributesW(path, NULL)) /* Check the existence */
+            return FALSE; /* Not found */
+    }
+
 #if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:786]: This exists also on Win2k3 SP2.
Also all this block (surrounded by the #if ...) seems to be done only if `flags & PRF_VERIFYEXISTS` is TRUE.

File: dll/win32/shell32/wine/shellpath.c [hunk 12]
```diff
@@ -713,41 +754,37 @@ BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
             if (!PathIsAbsoluteW(path))
                 return PathMakeAbsoluteW(path) && PathFileExistsAndAttributesW(path, NULL);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:757]:
```suggestion
            {
                if (!PathMakeAbsoluteW(path))
                    return FALSE;
                return PathFileExistsAndAttributesW(path, NULL);
            }
```

File: dll/win32/shell32/wine/shellpath.c [hunk 13]
```diff
@@ -713,41 +754,37 @@ BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
...
+    }
+
 #if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
-        if (flags & PRF_REQUIREABSOLUTE)
-        {
-            if (!PathIsAbsoluteW(path))
-                return PathMakeAbsoluteW(path) && PathFileExistsAndAttributesW(path, NULL);
-        }
+    if ((flags & PRF_REQUIREABSOLUTE) && !PathIsAbsoluteW(path))
+        return PathMakeAbsoluteW(path) && PathFileExistsAndAttributesW(path, NULL);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:788]:
```suggestion
    {
        if (!PathMakeAbsoluteW(path))
            return FALSE;
        return PathFileExistsAndAttributesW(path, NULL);
    }
```

File: sdk/include/reactos/shlwapi_undoc.h [hunk 14]
```diff
@@ -168,6 +168,7 @@ ShellMessageBoxWrapW(
 BOOL WINAPI PathFileExistsDefExtW(LPWSTR lpszPath, DWORD dwWhich);
 BOOL WINAPI PathFindOnPathExW(LPWSTR lpszFile, LPCWSTR *lppszOtherDirs, DWORD dwWhich);
 VOID WINAPI FixSlashesAndColonW(LPWSTR);
+BOOL WINAPI PathIsValidCharW(WCHAR, DWORD);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/include/reactos/shlwapi_undoc.h:171]: `class` is C++ keyword.

File: dll/win32/shell32/wine/shellpath.c (hunk 10 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: Used `_WIN32_WINNT_WS03`.

File: dll/win32/shell32/wine/shellpath.c (hunk 11 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: Used `_WIN32_WINNT_WS03`.
Moved code into `if (flags & PRF_VERIFYEXISTS)` block.

File: dll/win32/shell32/wine/shellpath.c [hunk 15]
```diff
@@ -146,90 +143,134 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+
+            /* If a '..' or '.' was encountered, skip to next component */
+            if (bDots)
+            {
+                while (*pchTemp && *pchTemp != L'\\')
+                {
+                    ++pchTemp;
+                }
+
+                if (*pchTemp == L'\\')
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:233]: Shouldn't this be a while-loop to collapse any consecutive backslashes?

File: sdk/include/reactos/shlwapi_undoc.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/include/reactos/shlwapi_undoc.h:171]: ok maybe "dwClass" then ^^
(I took these names from https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shlwapi/path.c;hb=e944dfa75fc49c297446eef2d3eda1a3eb59793e#l4416 , which is obviously in C ^^ )

File: dll/win32/shell32/wine/shellpath.c [hunk 16]
```diff
@@ -146,90 +143,134 @@ static BOOL WINAPI PathMakeAbsoluteW(LPWSTR path)
...
+                iDrive = PathGetDriveNumberW(szRoot);
+                if (iDrive != -1)
+                    PathBuildRootW(szRoot, iDrive); /* 'C:\' */
+            }
+
+            pchTemp = szTemp;
+            if (pchTemp[0] == L'\\')
+                ++pchTemp; /* The '\' only is a root */
+
+            StringCchCopyW(pszPath, MAX_PATH, szRoot);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:210]: Which 3 lines?

[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/shell32/wine/shellpath.c:210]: These ones:
```c
            pchTemp = szTemp;
            if (pchTemp[0] == L'\\')
                ++pchTemp; /* The '\' only is a root */
```

File: dll/win32/shell32/wine/shellpath.c [hunk 17]
```diff
@@ -673,81 +732,89 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
             if (!(flags & PRF_REQUIREABSOLUTE))
                 return TRUE;
 
             if (!PathIsAbsoluteW(path))
-                return PathMakeAbsoluteW(path) && PathFileExistsAndAttributesW(path, NULL);
+            {
+                if (!PathMakeAbsoluteW(path))
+                    return FALSE;
+                return PathFileExistsAndAttributesW(path, NULL);
+            }
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:779]: This #if-block could be made to look like the other one:
```suggestion
            if ((flags & PRF_REQUIREABSOLUTE) && !PathIsAbsoluteW(path))
            {
                if (!PathMakeAbsoluteW(path))
                    return FALSE;
                return PathFileExistsAndAttributesW(path, NULL);
            }
```
and replacing the #else below by #endif (and removing the old #endif)

File: dll/win32/shell32/wine/shellpath.c [hunk 18]
```diff
@@ -673,81 +732,89 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
         return TRUE;
     }
-    else if (PathIsFileSpecW(path))
+
+    if (PathIsFileSpecW(path)) /* Filename only */
     {
-        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
-            return TRUE;
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && /* Try to find a program? */
+            PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich)) /* Search it */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:763]:
```suggestion
        /* Try to find the path with program extensions applied? */
        if ((flags & PRF_TRYPROGRAMEXTENSIONS) &&
            PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
```

File: dll/win32/shell32/wine/shellpath.c [hunk 19]
```diff
@@ -673,81 +732,89 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
+
+    /* Qualify the path */
+    PathQualifyExW(path, ((flags & PRF_FIRSTDIRDEF) ? *dirs : NULL), 1);
+
+    TRACE("(%s)\n", debugstr_w(path));
+
+    if (flags & PRF_VERIFYEXISTS) /* Verify the existence? */
+    {
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && /* Try to find a program? */
+            PathSearchOnExtensionsW(path, dirs, FALSE, dwWhich)) /* Search it */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:799]:
```suggestion
        /* Try to find the path with program extensions applied? */
        if ((flags & PRF_TRYPROGRAMEXTENSIONS) &&
            PathSearchOnExtensionsW(path, dirs, FALSE, dwWhich))
```

File: dll/win32/shell32/wine/shellpath.c [hunk 20]
```diff
@@ -673,81 +732,89 @@ BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
...
-        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
-            return TRUE;
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && /* Try to find a program? */
+            PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich)) /* Search it */
+        {
+            return TRUE; /* Found */
+        }
 
-        if (PathFindOnPathW(path, dirs))
+        if (PathFindOnPathW(path, dirs)) /* Try to find the filename in the directories */
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:768]:
```suggestion
        /* Try to find the filename in the directories */
        if (PathFindOnPathW(path, dirs))
```

**Review by HBelusca — APPROVED** [review]

**Comment by katahiromz** [comment]
errors
112 failures. We have embugged at somewhere.

**Comment by katahiromz** [comment]
passed
Now tests passed.
